src/server/game/Entities/Object/Object.cpp

```diff
@@ -1428,9 +1428,15 @@ void WorldObject::UpdateAllowedPositionZ(float x, float y, float &z) const
         }
         else
         {
-            float ground_z = GetMapHeight(x, y, z) + unit->GetHoverOffset();
-            if (z < ground_z)
-                z = ground_z;
+            float const gridHeight = GetMap()->GetGridHeight(x, y) + unit->GetHoverOffset();
+            float const vmapHeight = GetMap()->GetVmapDataFloorZ(x, y, z); // don't add hover offset here. if check might return unintended value
+            // Check if vmap under us
+            if (vmapHeight <= INVALID_HEIGHT)
```

Comment:
Why did you split these calls? GetMapHeight uses GetMap()->GetHeight which is supposed to return the correct height value, either from grid or vmaps

Reply:
The changes made in Map::GetHeight will make it so that it will select the closest one, not caring if z is lower than grid. The drakes in the caverns of time will rise up again if I use that function.

I'm thinking about reverting that part so we can still use the same function but not sure yet if it could break things again